src/consumer/runner.js

```diff
@@ -297,24 +299,63 @@ module.exports = class Runner {
     }
 
     const concurrently = limitConcurrency({ limit: this.partitionsConsumedConcurrently })
-    await Promise.all(
-      batches.map(batch =>
-        concurrently(async () => {
-          if (!this.running) {
-            return
-          }
-
-          if (batch.isEmpty()) {
-            return
-          }
-
-          await onBatch(batch)
-        })
-      )
-    )
 
-    await this.autoCommitOffsets()
-    await this.consumerGroup.heartbeat({ interval: this.heartbeatInterval })
+    if (!this.running) {
+      return
+    }
+
+    const enqueuedTasks = []
+    let expectednumberOfExecutions = 0
+    let numberOfExecutions = 0
+    const { lock, unlock, unlockWithError } = barrier()
+
+    while (true) {
+      const result = iterator.next()
+      if (result.done) {
+        break
+      }
+
+      enqueuedTasks.push(async () => {
+        if (!this.running) {
+          return
+        }
+
+        const batches = await result.value
+        expectednumberOfExecutions += batches.length
+
+        batches.map(batch =>
+          concurrently(async () => {
+            try {
+              if (!this.running) {
+                return
+              }
+
+              if (batch.isEmpty()) {
+                return
+              }
+
+              await onBatch(batch)
+              await this.autoCommitOffsets()
```

Comment:
This changes the behavior compared to before. Unless I'm mistaken, before we would only autoCommitOffsets and heartbeat after all the batches, not in between.

Reply:
you are right, but I think we should have the hearbeats here, since they won't trigger if the interval is not right.

Reply:
9c24622